### PR TITLE
Support of byUTF for ubyte[] argument

### PR DESCRIPTION
**Pull request by vporton**
For the reasons outlined in the discussion of [that pull request](https://github.com/dlang/phobos/pull/7247), we concluded that we need to be able to call `byUTF` on the argument of type `ubyte[]`. This PR implements exactly that.

I remind that this is necessary:

1. to support converting a chunk extracted from a file;
1. to eliminate the need to validate a string of `char`s two times: when it is created and when converted by `byUTF` (this simplifies programming and improves efficiency).

### CODE REVIEW

**Comment by lesderid** [comment]
What happens for a range of `ubyte`s that aren't valid UTF-8? Is this covered by a test?

File: std/utf.d [hunk 1]
```diff
@@ -4354,6 +4354,9 @@ if (isSomeChar!C)
     // hellö as a range of `ubyte`s, which are UTF-8
     assert((cast(ubyte[]) [0x68, 0x65, 0x6c, 0x6c, 0xC3, 0xB6]).byUTF!char().equal(['h', 'e', 'l', 'l', 0xC3, 0xB6]));
 
+    assertThrown((cast(ubyte[]) [0xC3, 0x28]).byUTF!char());
```
Comment (lesderid) [inline std/utf.d]: Needs an `import std.exception : assertThrown;`.

File: std/utf.d [hunk 2]
```diff
@@ -60,7 +60,7 @@ $(TR $(TD Miscellaneous) $(TD
    +/
 module std.utf;
 
-import std.exception;  // basicExceptionCtors
+import std.exception;  // basicExceptionCtors, assertThrown
```
Comment (lesderid) [inline std/utf.d]: This won't work because tests are extracted. Please import it inside the unittest block.

**Comment by atilaneves** [comment]
I read the discussion on the other PR and AIUI it called for something that took `ubyte[]` and lazily produced validaded `char[]`. This doesn't seem to be it. Could you please explain what you're trying to accomplish? Thanks.

File: std/utf.d [hunk 3]
```diff
@@ -4331,16 +4338,26 @@ if (isSomeChar!C)
...
     // 𐐷 is four code units in UTF-8, two in UTF-16, and one in UTF-32
-    "𐐷".byUTF!char().equal([0xF0, 0x90, 0x90, 0xB7]);
-    "𐐷".byUTF!wchar().equal([0xD801, 0xDC37]);
-    "𐐷".byUTF!dchar().equal([0x00010437]);
+    assert("𐐷".byUTF!char().equal([0xF0, 0x90, 0x90, 0xB7]));
+    assert("𐐷".byUTF!wchar().equal([0xD801, 0xDC37]));
+    assert("𐐷".byUTF!dchar().equal([0x00010437]));
+
+    // hellö as a range of `ubyte`s, which are UTF-8
+    assert((cast(ubyte[]) [0x68, 0x65, 0x6c, 0x6c, 0xC3, 0xB6]).byUTF!char().equal(['h', 'e', 'l', 'l', 0xC3, 0xB6]));
```
Comment (dukc) [inline std/utf.d:4356]: Good, but I think you should test (directly) converting an `ubyte` range to utf-16 and/or utf-32

**Comment by dukc** [comment]
> I read the discussion on the other PR and AIUI it called for something that took `ubyte[]` and lazily produced validaded `char[]`. This doesn't seem to be it. Could you please explain what you're trying to accomplish? Thanks.

The added unittest explains it:
`assert((cast(ubyte[]) [0x68, 0x65, 0x6c, 0x6c, 0xC3, 0xB6]).byUTF!char().equal(['h', 'e', 'l', 'l', 0xC3, 0xB6]));`

You pass in a range of `ubyte` and get a range of `char`. Handy, as no separate step to deal with autodecoding is needed.

File: std/utf.d (hunk 3 above)
Comment (dukc) [reply std/utf.d:4356]: And also dealing with invalid UTF-8 in the `ubyte`, both with exceptions and replacement characters.

**Comment by dukc** [comment]
Perhaps this should also accept `ushort` (assumed to be UTF-16) and `uint` (UTF-32)?

Of course, this can be implemented later on just as well.

**Comment by RazvanN7** [comment]
ping @vporton

**Comment by LightBender** [comment]
PR closed as stalled. This can be resurrected for Phobos 3.

**Pull request closed without merging**